pkg/store/store.go

```diff
@@ -17,44 +17,170 @@ limitations under the License.
+// adding an RG requires looking up all applicable originID's via the index and allowing their references
+type grantGraph map[subject]map[targetResourceGroup]map[CRPID]map[originResourceID]map[targetNamespacedName]map[purpose]grant
+
+type namespace string
+type originResourceGroup schema.GroupResource
+
+// originResourceIndex is used to find applicable grants to add/remove a ReferenceGrant ID for the grantors list
+type originResourceIndex map[namespace]map[originResourceGroup]originResourceID
+
 // in-memory GraphStore
 type AuthStore struct {
-	graph map[string]map[string]bool
-	mutex sync.RWMutex
+	graph               grantGraph
+	originResourceIndex originResourceIndex
+	mutex               sync.RWMutex
 }
 
 func NewGraphStore() *AuthStore {
-	return &AuthStore{graph: make(map[string]map[string]bool)}
+	return &AuthStore{graph: make(grantGraph)}
 }
 
 func (s *AuthStore) CheckAuthz(sar authorizationv1.SubjectAccessReview) (bool, error) {
 	s.mutex.RLock()
 	defer s.mutex.RUnlock()
-	resource := sar.Spec.ResourceAttributes.Resource
-	verb := sar.Spec.ResourceAttributes.Verb
 
-	if verbs, ok := s.graph[resource]; ok {
-		if allowed, ok := verbs[verb]; ok {
-			return allowed, nil
+	user := sar.Spec.User
+	groups := sar.Spec.Groups
+	trg := targetResourceGroup{
+		Resource: sar.Spec.ResourceAttributes.Resource,
+		Group:    sar.Spec.ResourceAttributes.Group,
+	}
+	orig := originResourceID(strings.Join(sar.Spec.Extra[ExtraOriginKey], ", ")) // TODO: where can we get this value from the client request properly?
+	nn := targetNamespacedName{
+		namespace: sar.GetName(),
+		name:      sar.GetNamespace(),
+	}
+	p := purpose(strings.Join(sar.Spec.Extra[ExtraPurposeKey], ", ")) // TODO: where can we get this value from the client request properly?
+	v := verb(sar.Spec.ResourceAttributes.Verb)
+
+	var hasErr error
+	for _, g := range groups {
+		allowed, err := s.lookup(subject(g), trg, orig, nn, p, v)
+		if err != nil {
+			hasErr = err // TODO: multi error
+		}
+		if allowed {
+			return true, hasErr
 		}
 	}
+	allowed, err := s.lookup(subject(user), trg, orig, nn, p, v)
+	if err != nil {
+		hasErr = err // TODO: multi error
+	}
+	return allowed, hasErr
+}
+
+func (s *AuthStore) lookup(subj subject, trg targetResourceGroup, orig originResourceID, nn targetNamespacedName, p purpose, v verb) (bool, error) {
+	// if verbs, ok := s.graph[targetResource]; ok {
+	// 	if allowed, ok := verbs[verb]; ok {
+	// 		return allowed, nil
+	// 	}
+	// }
+
+	trgMap, ok := s.graph[subj]
+	if !ok {
+		return false, nil
+	}
+	crpMap, ok := trgMap[trg]
+	if !ok {
+		return false, nil
+	}
+	// lookup from every possible CRP for this targetResourceGroup
+	for crp, origMap := range crpMap {
+		if orig == "" {
+			// no originResourceID provided, fallback and check for any possibilities
+			for o, nnMap := range origMap {
+				if pgMap, ok := nnMap[nn]; ok {
+					if checkGrant(pgMap, p, v) {
+						return true, nil
+					}
```

Comment:
Ah, I figured this would make the control-flow more complicated.
You are right -- In the same-namespace case, access will be just as fast as RBAC if we do that.
However, it is not without trade-off.
Consider TLS Secrets.
If we open access for all Secrets in a Namespace, it will include application API keys, database passwords, etc.
Restricting to only referenced names in the same-namespace case still has a security benefits that come with the cost of the graph update latency and storage space.